File: imu_calib_python/imu_calib/tasks/run_gyro_bias.py
from __future__ import annotations

from dataclasses import is_dataclass

from imu_calib.calib.estimate_gyro_bias import estimate_gyro_bias
from imu_calib.tasks.make_task_result import make_task_result
from imu_calib.tasks.require_inputs import require_inputs


def run_gyro_bias(input_data) -> dict:
    static = _normalize_static_input(input_data)
    missing = require_inputs(static, ["gyro"])
    if missing:
        return make_task_result(
            False,
            "static.gyro is required for gyro bias estimation.",
            None,
            missing_inputs=["static.gyro"],
            meta={"task_name": "run_gyro_bias"},
        )

    bg, info = estimate_gyro_bias(static.gyro, t=getattr(static, "t", None), temp=getattr(static, "temp", None))
    return make_task_result(
        True,
        "gyro bias estimated successfully.",
        {"bg": bg, "stats": info},
        meta={"task_name": "run_gyro_bias"},
    )


def _normalize_static_input(input_data):
    if is_dataclass(input_data) and hasattr(input_data, "static"):
        return input_data.static
    if isinstance(input_data, dict) and "static" in input_data:
        return input_data["static"]
    return input_data

File: imu_calib_matlab/src/tasks/run_gyro_bias.m
function task = run_gyro_bias(inputData)
%RUN_GYRO_BIAS Run gyroscope static bias estimation using the minimum required input.
% Minimum input:
%   static.gyro
% Optional:
%   static.t, static.temp

staticData = normalize_static_input(inputData);
missing = require_inputs(staticData, {'gyro'});
if ~isempty(missing)
    task = make_task_result(false, 'static.gyro is required for gyro bias estimation.', [], ...
        'missing_inputs', {'static.gyro'}, ...
        'meta', struct('task_name', 'run_gyro_bias'));
    return;
end

t = getfield_or_default(staticData, 't', []); %#ok<GFLD>
temp = getfield_or_default(staticData, 'temp', []); %#ok<GFLD>
[bg, info] = estimate_gyro_bias(staticData.gyro, 't', t, 'temp', temp);

result = struct();
result.bg = bg;
result.stats = info;

task = make_task_result(true, 'gyro bias estimated successfully.', result, ...
    'meta', struct('task_name', 'run_gyro_bias'));
end

function staticData = normalize_static_input(inputData)
if isstruct(inputData) && isfield(inputData, 'static')
    staticData = inputData.static;
else
    staticData = inputData;
end
end

function value = getfield_or_default(S, fieldName, defaultValue)
if isstruct(S) && isfield(S, fieldName)
    value = S.(fieldName);
else
    value = defaultValue;
end
end
